### PR TITLE
DAOS-6879 object: Select I/O RDG by JCH

### PR DESCRIPTION
**Pull request by gnailzenh**
Use jump consistent hash (JCH) to select redundancy group for
I/O, so sequential I/O can be randomized and balance workload.
JCH could also be better for supporting progressive layout in
the future.

Signed-off-by: Liang Zhen <liang.zhen@intel.com>

### CODE REVIEW

**Review by jolivier23 — DISMISSED** [review]

File: src/object/cli_obj.c [hunk 1]
```diff
@@ -620,8 +620,7 @@ obj_dkey2grpidx(struct dc_object *obj, uint64_t hash, unsigned int map_ver)
 
 	D_ASSERT(obj->cob_shards_nr >= grp_size);
 
-	/* XXX, consistent hash? */
-	grp_idx = hash % (obj->cob_shards_nr / grp_size);
+	grp_idx = d_hash_jump(hash, obj->cob_shards_nr / grp_size);
```
Comment (wangdi1) [inline src/object/cli_obj.c:623]: Will JCH make sure the integer dkey evenly distributed over shards? probably fine.

**Review by wangdi1 — DISMISSED** [review]

**Comment by ashleypittman** [comment]
These failures appear to be an NLT bug triggered by readdir (and therefore object iterate) returning entries in a different order, I'm working on a fix.

**Comment by jolivier23** [comment]
Updated to latest master after landing @ashleypittman's patch

File: src/object/cli_obj.c (hunk 1 above)
Comment (gnailzenh) [reply src/object/cli_obj.c:623]: Yes, it does, it's better than (murmur_hash(integer) % buckets)

**Comment by gnailzenh** [comment]
Vish confirmed this patch has similar performance as hashed integer

File: src/object/cli_obj.c (hunk 1 above)
Comment (wangdi1) [reply src/object/cli_obj.c:623]: Actually I mean compared with offset % buckets (original integer dkey with DFS), which is even completely.
Interesting we do similar for lustre, why there is no complain there.

**Review by vvenkates27 — DISMISSED** [review]

**Comment by gnailzenh** [comment]
I've added a performance test to this patch

File: src/gurt/tests/test_gurt.c [hunk 2]
```diff
@@ -1990,6 +1990,101 @@ test_gurt_string_buffer(void **state)
...
+			bkt_min = counters[i];
+	}
+	stdiv = d_stand_div(counters, buckets);
+	duration = (double)d_timediff_ns(&then, &now) / NSEC_PER_SEC;
+
+	fprintf(stdout,
+		"Hash: %s, bkts: %d, min/max: %d/%d, "
+		"range: %d, stdiv: %F, rate: %F\n",
+		hash2name(hash_type), buckets, (int)bkt_min, (int)bkt_max,
+		(int)(bkt_max - bkt_min), stdiv, (double)loop / duration);
```
Comment (jolivier23) [inline src/gurt/tests/test_gurt.c:2070]: should free the counters

**Review by johannlombardi — APPROVED** [review]

**Pull request merged**